[10.0] Add modules agreement_account and agreement_sale

alexis-via opened the pull request:
See README.rst of the modules for more info. This PR is the result of a discusson that I started in the contributors mailing list on July 13th 2017 with subject "Contract without analytic".

alexis-via committed:
Add modules agreement_account and agreement_sale

alexis-via committed:
Copy sale_agreement_id from invoice to refund

bealdav commented on agreement_account/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': u'Agreement (Account)',
```
`u` required ?

bealdav commented:
Nice work for me. Thanks Alexis :+1: 
Just a simple remark: on my side I think we could name the module `agreement` and the main model `agreement.agreement`. Then we could use it sale side and purchase side.

bealdav commented:
ping @elicoidal @jgrandguillaume

alexis-via committed:
Rename 'sale.agreement' model to 'agreement'

Add type field on 'agreement': 'sale' or 'purchase'
Add agreement_id field on supplier invoices

alexis-via commented:
@bealdav I took into account your remark: I renamed the model 'sale.agreement' to 'agreement'. Therefore, I also added a 'type' field (sale or purchase) on it and the field is now also displayed on supplier invoices.
I also updated all the modules impacted by this change in OCA/edi and OCA/l10n-france... it was not a small change !

pedrobaeza commented:
Please don't call any field `type`, as you are overwriting Python keyword. Call it `agreement_type` for example.

bealdav commented:
@pedrobaeza another option is to name `type_`

alexis-via committed:
Rename field type to agreement_type

Small fix in view

rousseldenis commented:
LGTM. 

@alexis-via  I ask myself if this could interfere with this :

https://github.com/odoo/odoo/blob/875dd9221f80ea1b60cd9ee8a005d71246734a2b/addons/purchase_requisition/models/purchase_requisition.py#L9

rousseldenis commented on agreement_account/models/agreement.py:
```diff
@@ -0,0 +1,49 @@
+    _description = 'Agreement'
+
+    code = fields.Char(
+        string='Code', required=True, copy=False)
```
string is not needed anymore in that case (and below too).

rousseldenis commented on agreement_account/models/agreement.py:
```diff
@@ -0,0 +1,49 @@
+        'account.invoice', 'agreement_id', string='Supplier Invoices',
+        readonly=True, domain=[('type', 'in', ('in_invoice', 'in_refund'))])
+
+    def name_get(self):
```
Must decorate api.multi

bealdav replied:
As far as I know all methods which inherit of models.Model are `api.multi` by default in v10 (dixit Antony Lesuisse @ last odoo days). Then we should only specify decorator when don't want this behavior.
It allows to making the code lighter. 
Sure it was different in previous versions.

rousseldenis reviewed (changes requested):
Code review. Minor comments.

Could we add maybe a date range for agreement validity (not required field)?

Could you add tests ?

alexis-via committed:
Remove unneeded string argument in field definition

alexis-via commented:
There are certainly many fields that could be added on the "agreement" object. It will depend very much on what you will use it for. I only added the very minimum here. I don't mind adding a dependancy on date_range and add such an optional field, but there are so many that can be added too... I tried to keep it minimal.
About tests, as you can see, these modules contain almost no code... so there is no real interesting things to test. The main objective of these modules is to store information that we need for electronic invoices.
